### PR TITLE
Improve memory pool performance

### PR DESCRIPTION
**Pull request by okuta**
This PR reduces 20-30% memory allocation time. [Benchmark code](https://github.com/okuta/cupy-bench)
Please merge #1540 first.

before
```
% python alloc_empty.py
alloc, 2.559044, 2.557237
alloc_and_free, 3.869997, 3.868559

% python alloc.py
dummy_call, 0.000104, 0.000099
mem_0B, 0.000735, 0.000728
mem_1K, 0.001575, 0.001568
mem_1M, 0.001610, 0.001602
```

after
```
% python alloc_empty.py
alloc, 2.061816, 2.060394
alloc_and_free, 3.069911, 3.068501

% python alloc.py
dummy_call, 0.000107, 0.000102
mem_0B, 0.000647, 0.000641
mem_1K, 0.001154, 0.001147
mem_1M, 0.001180, 0.001174
```

### CODE REVIEW

**Comment by okuta** [comment]
jenkins, test this please.

File: cupy/cuda/memory.pyx [hunk 1]
```diff
@@ -701,171 +805,103 @@ cdef class SingleDeviceMemoryPool:
...
-                arena_index.insert(arena_index.begin() + index, bin_index)
-                arena.insert(index, free_list)
-            free_list.add(chunk)
-        finally:
-            rlock.unlock_fastrlock(self._free_lock)
-
-    cpdef bint _remove_from_free_list(self, Py_ssize_t size, chunk,
-                                      size_t stream_ptr) except *:
-        """Removes the chunk from the free list.
+    cdef vector.vector[int8_t]* _arena_count(self, size_t stream_ptr):
```
Comment (beam2d) [inline cupy/cuda/memory.pyx]: This function looks always returning 0 or 1. Is my understanding correct? (just want to confirm my understanding)

Comment (okuta) [reply cupy/cuda/memory.pyx]: No. This function is a No. This function is a same as `_arena` and `_arena_index`.same as `_arena` and `_arena_index`.

**Comment by okuta** [comment]
I rebase.
jenkins, test this please.

**Comment by okuta** [comment]
I changed local variable name

**Pull request merged**